src/Storages/Statistics/StatisticsTDigest.cpp

```diff
@@ -43,31 +41,35 @@ void StatisticsTDigest::deserialize(ReadBuffer & buf)
 
 Float64 StatisticsTDigest::estimateLess(const Field & val) const
 {
-    auto val_as_float = StatisticsUtils::tryConvertToFloat64(val);
-    if (val_as_float)
-        return t_digest.getCountLessThan(*val_as_float);
-    throw Exception(ErrorCodes::LOGICAL_ERROR, "Statistics 'tdigest' does not support estimating value of type {}", val.getTypeName());
+    Field val_converted = convertFieldToType(val, *data_type);
```

Comment:
What will happen if `val < '10.5'` but val is int, will `< 10.5` become `< 10`?

Reply:
or `val < '9999999999999999999999999'` but val is int16

Reply:
1. `val_int32 < '10.5'`: the conversion will fail,  but it will never be executed, because the query will fail in the quey plan planning stage which is prior to the optimization stage.
2. `val_int32 < 10.5`: the conversion will get `Null`, which I have fixed here https://github.com/ClickHouse/ClickHouse/pull/67013/commits/aafe498b7f6045e75414c17abecad168aa0efb88 and https://github.com/ClickHouse/ClickHouse/pull/67013/commits/cfaa852af9e94bd25217bfbb690f2d6ad641f3d6
3. `val_int16 < '9999999999999999999999999'` the conversion will get `Null`, also fixed in the previous commits.

Reply:
Thanks. I now included the other two commits in this PR.

(Side note: with that, `StatisticsUtils::tryConvertToFloat64` now contains some random-looking special cases, which makes it a time bomb. We really really need to rewrite the tests to check all edge cases systematically, see https://github.com/ClickHouse/ClickHouse/pull/67013#discussion_r1703151496)

Reply:
That's why I tried to introduce `TargetType`, but this looks okay to me